language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -0,0 +1,150 @@
+still have to be filled in. For builtins (e.g. +, etc.) the cost tracking can be
+a bit more complicated, as the required resources may depend on the size of the
+inputs (E.g. multiplying numbers, where the output will be around 6 words if
+both inputs are at 3 words each). These cost estimations will also have factors
+attached which can be configured at runtime via protocol parameters - so it's
+possible to adjust them at runtime.
+
+-}
+
+module Language.PlutusCore.Evaluation.Machine.ExBudgeting
+    ( CekBudgetMode(..)
+    , ExBudget(..)
+    , ExBudgetState(..)
+    , ExTally(..)
+    , ExTallyCounter(..)
+    , estimateStaticStagedCost
+    , exBudgetCPU
+    , exBudgetMemory
+    , exTallyMemory
+    , exTallyCPU
+    , exBudgetStateBudget
+    , exBudgetStateTally
+    )
+where
+
+
+import           Language.PlutusCore
+import           PlutusPrelude
+
+import           Control.Lens.Indexed
+import           Control.Lens.TH                                         (makeLenses)
+import           Data.HashMap.Monoidal
+import           Data.List                                               (intersperse)
+import           Data.Text.Prettyprint.Doc
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Machine.GenericSemigroup
+
+data CekBudgetMode =
+      Counting -- ^ For precalculation
+    | Restricting -- ^ For execution, to avoid overruns
+
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+    deriving (Eq, Show, Generic)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExBudget)
+instance PrettyBy config ExBudget where
+    prettyBy config (ExBudget cpu memory) = parens $ fold
+        [ "{ cpu: ", prettyBy config cpu, line
+        , "| mem: ", prettyBy config memory, line
+        , "}"
+        ]
+
+data ExBudgetState = ExBudgetState
+    { _exBudgetStateTally  :: ExTally -- ^ for counting what cost how much
+    , _exBudgetStateBudget :: ExBudget  -- ^ for making sure we don't spend too much
+    }
+    deriving stock (Eq, Generic, Show)
+instance ( PrettyBy config (Term TyName Name ())
+         ) => PrettyBy config ExBudgetState where
+    prettyBy config (ExBudgetState tally budget) = parens $ fold
+        [ "{ tally: ", prettyBy config tally, line
+        , "| budget: ", prettyBy config budget, line
+        , "}"
+        ]
+
+data ExTally = ExTally
+    { _exTallyCPU    :: ExTallyCounter ExCPU
+    , _exTallyMemory :: ExTallyCounter ExMemory
+    }
+    deriving stock (Eq, Generic, Show)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+instance ( PrettyBy config (Term TyName Name ())
+         ) => PrettyBy config ExTally where
+    prettyBy config (ExTally cpu memory) = parens $ fold
+        [ "{ cpu: ", prettyBy config cpu, line
+        , "| mem: ", prettyBy config memory, line
+        , "}"
+        ]
+
+newtype ExTallyCounter unit = ExTallyCounter (MonoidalHashMap (Plain Term) [unit])
```

Comment:
It strikes me that I'm not sure this is tallying the right thing. It's indexing the usage by the *term*. But how often is the very same term going to come up? I think what I thought we were going to index by was the *type of thing that incurred the cost*, i.e. all the calls to `spendBoth` in `Cek.hs` would have their own tag, and that would let us see that e.g. we're spending lots of money on wrap/unwrap steps or something. Which seems more useful than this.

Reply:
The keys for the tally would presumably also be the keys for the simple bit (i.e. the bit that isn't worrying about builtin functions) of the cost model when we do that.

Reply:
> It's indexing the usage by the term.

I presume we don't want this to happen on-chain? It looks like this costing machinery might consume more memory and CPU than evaluation itself. Which might be another argument in favor of abstracting things out.

Reply:
If we have a count per *type* of operation, it should only cost a fixed amount proportional to the number of types, not much worse than the overall budget which we already track.

Reply:
The cost difference between indexing by term and by type is mostly the hashing of the term. But yeah, both versions are somewhat expensive. I think I explicitly removed the ability to disable the functionality a while back.

Reply:
> If we have a count per type of operation, it should only cost a fixed amount proportional to the number of types, not much worse than the overall budget which we already track.

Sure, I mean the _current_ costing machinery.

Reply:
> The cost difference between indexing by term and by type is mostly the hashing of the term.

I don't think so - there are *way* more terms, and they're not that interesting to group by.

To be clear, when I say "the type of thing that incurred the cost", I mean things like "wrap/unwrap" or "entered function application frame".